linchpin/__init__.py

```diff
@@ -73,10 +72,8 @@ def _handle_results(ctx, results):
                     msg = tr._check_key('msg')
                     ctx.log_state("Target '{0}': {1} failed with"
                                   " error '{2}'".format(k, tr._task, msg))
-                    sys.exit(retval)
-        else:
-            if v:
-                sys.exit(v)
+
+    sys.exit(return_code)
 
```

Comment:
There are scenarios where performing a sys.exit here is not the desired case. Adjusting to this means linchpin exits here no matter what. If that's the case, then there is no reason to perform exits elsewhere in the code. For instance, if the `return_code` is 0, but there will be an error provisioning only one node of 3, this may not error out in the future Linchpin may print this out and continue on through the process.

Because of some code,  I recently had to update this as well. Here's how I implemented the changes. I'm not sure they are perfect yet, but they work well.

```
def _handle_results(ctx, results, return_code=1):
    ..snip..
    for k, v in results.iteritems():
        if not isinstance(v, int):
            trs = v

            if trs is not None:
                trs.reverse()
                tr = trs[0]
                if tr.is_failed():
                    msg = tr._check_key('msg')
                    ctx.log_state("Target '{0}': {1} failed with"
                                  " error '{2}'".format(k, tr._task, msg))
                    sys.exit(return_code)
        else:
            if v:
                sys.exit(v)
```
One thing to consider when updating an existing function is that adding a new argument may affect callers from other classes, functions, etc. In that case, it's a good idea to make it a kwarg and default it to some value. In that way, it doesn't affect other callers that haven't updated the function.

Additionally, I'm curious why you remove the last 3 lines. Maybe that's why it's not returning the proper error code? In those cases though, the function should only return errors, I believe. Can you verify?